src/libcore/iter/traits.rs

```diff
@@ -519,6 +519,31 @@ pub trait ExactSizeIterator: Iterator {
         assert_eq!(upper, Some(lower));
         lower
     }
+
+    ///
+    /// Returns whether the iterator is empty.
+    ///
+    /// This method has a default implementation using `self.len()`, so you
+    /// don't need to implement it yourself.
+    ///
+    /// # Examples
+    ///
+    /// Basic usage:
+    ///
+    /// ```
+    /// let mut one_element = [0].iter();
+    /// assert!(!one_element.is_empty());
+    ///
+    /// assert_eq!(one_element.next(), Some(0));
+    /// assert!(one_element.is_empty());
+    ///
+    /// assert_eq!(one_element.next(), None);
+    /// ```
+    #[inline]
+    #[unstable(feature = "exact_size_is_empty", issue = "0")]
+    fn is_empty(&self) -> bool {
+        self.len() == 0
```

Comment:
Instead of calling `.len()` you could also only check `let (lower, upper) = self.size_hint(); upper == 0` and skip the assert. (I'm sure sure this is better.)